File: solutions/scala/133-binary-search.scala
/**
 * Problem 133: Binary Search (LeetCode 704)
 * Difficulty: Easy
 * Language: Scala
 */
def binarySearch(nums: Array[Int], target: Int): Int = {
  def search(lo: Int, hi: Int): Int = {
    if (lo > hi) -1
    else {
      val mid = lo + (hi - lo) / 2
      if (nums(mid) == target) mid
      else if (nums(mid) < target) search(mid + 1, hi)
      else search(lo, mid - 1)
    }
  }
  search(0, nums.length - 1)
}

File: solutions/elixir/133-binary-search.ex
# Problem 133: Binary Search (LeetCode 704)
# Difficulty: Easy
# Language: Elixir
# 
defmodule Solution do
  def binary_search(nums, target) do
    arr = :array.from_list(nums)
    search(arr, target, 0, :array.size(arr) - 1)
  end
  defp search(_, _, lo, hi) when lo > hi, do: -1
  defp search(arr, target, lo, hi) do
    mid = lo + div(hi - lo, 2)
    val = :array.get(mid, arr)
    cond do
      val == target -> mid
      val < target  -> search(arr, target, mid + 1, hi)
      true          -> search(arr, target, lo, mid - 1)
    end
  end
end
